agent/unit-test/softwarecontaineragent_unittest.cpp

```diff
@@ -0,0 +1,213 @@
+ * SOFTWARE.
+ *
+ * For further information see LICENSE
+ */
+
+#include "softwarecontaineragent.h"
+#include "softwarecontainererror.h"
+#include "config/config.h"
+#include "config/configloader.h"
+#include "config/mainconfigsource.h"
+#include "config/configdefinition.h"
+
+#include <gmock/gmock.h>
+#include <gtest/gtest.h>
+
+using namespace softwarecontainer;
+
+/*
+ * Test stub - StringConfigLoader
+ *
+ * Loads a Glib::KeyFile config from a string, compared to the "real" loader which reads
+ * from file.
+ *
+ * This stub is taken from softwarecontaineragent_componenttest
+ */
+class StringConfigLoader : public ConfigLoader
+{
+
+LOG_DECLARE_CLASS_CONTEXT("CFGL", "SoftwareContainer general config loader");
+
+public:
+    // Constructor just needs to init parent with the config source string
+    StringConfigLoader(const std::string &source) : ConfigLoader(source) {}
+
+    std::unique_ptr<Glib::KeyFile> loadConfig() override
+    {
+        std::unique_ptr<Glib::KeyFile> configData = std::unique_ptr<Glib::KeyFile>(new Glib::KeyFile);
+        try {
+            configData->load_from_data(Glib::ustring(this->m_source), Glib::KEY_FILE_NONE);
+        } catch (Glib::KeyFileError &error) {
+            log_error() << "Could not load SoftwareContainer config: \"" << error.what() << "\"";
+            throw error;
+        }
+
+        return configData;
+    }
+};
+
+/*
+class MockSCA :
+    public SoftwareContainerAgent
+{
+public:
+    MockSCA(Glib::RefPtr<Glib::MainContext> mainLoopContext,
+            std::shared_ptr<Config> config) :
+            SoftwareContainerAgent(mainLoopContext, config)
+    {
+
+    }
+
+    MOCK_METHOD2(doCreateHelper, void(ContainerID containerID, const SoftwareContainerConfig *config));
+    void createHelper(ContainerID containerID, std::unique_ptr<const SoftwareContainerConfig> config) {
+        doCreateHelper(containerID, config.get());
+    }
+};
+
+*/
+namespace softwarecontainer {
+class TestContainerInterface : public SoftwareContainerAbstractInterface
+{
+    LOG_DECLARE_CLASS_CONTEXT("SCAU", "SoftwareContainerAgent Unit Test");
+public:
+    TestContainerInterface() {};
+
+    MOCK_METHOD1(startGateways, ReturnCode(const GatewayConfiguration&));
+
+    MOCK_METHOD1(createCommandJob, std::shared_ptr<CommandJob>(const std::string&));
+
+    MOCK_METHOD1(shutdown, void(unsigned int));
+
+    MOCK_METHOD0(suspend, void());
+
+    MOCK_METHOD0(resume, void());
+
+    MOCK_METHOD3(bindMount, ReturnCode(const std::string &, const std::string &, bool readonly));
+
+    bool previouslyConfigured()
+    {
+        return false;
+    }
+
+};
+
+class TestFactory: public SoftwareContainerFactory
+{
+public:
+    TestFactory(std::shared_ptr<SoftwareContainerAbstractInterface> testContainerInterface) :
+        m_container(testContainerInterface)
+    { }
+
+    std::shared_ptr<SoftwareContainerAbstractInterface> createContainer(const ContainerID id __attribute__((unused)),
```

Comment:
I generally don't like the `__attribute__` attributes, I would rather have a compiler warning actually.